src/data_device_manager/mod.rs

```diff
@@ -0,0 +1,333 @@
+pub mod data_device;
+pub mod data_offer;
+pub mod data_source;
+
+use std::{
+    fs, io,
+    os::unix::prelude::{AsRawFd, FromRawFd, IntoRawFd, RawFd},
+};
+
+use wayland_backend::io_lifetimes::{AsFd, FromFd, OwnedFd};
+use wayland_client::{
+    globals::{BindError, GlobalList},
+    protocol::{
+        wl_data_device,
+        wl_data_device_manager::{self, DndAction, WlDataDeviceManager},
+        wl_data_source::WlDataSource,
+        wl_seat::WlSeat,
+    },
+    Connection, Dispatch, QueueHandle,
+};
+
+use crate::{
+    error::GlobalError,
+    globals::{GlobalData, ProvidesBoundGlobal},
+};
+
+use self::{
+    data_device::{DataDeviceData, DataDeviceDataExt},
+    data_source::{CopyPasteSource, DataSourceData, DataSourceDataExt, DragSource},
+};
+
+#[derive(Debug)]
+pub struct DataDeviceManagerState {
+    manager: WlDataDeviceManager,
+}
+
+impl DataDeviceManagerState {
+    pub fn bind<State>(globals: &GlobalList, qh: &QueueHandle<State>) -> Result<Self, BindError>
+    where
+        State: Dispatch<WlDataDeviceManager, GlobalData, State> + 'static,
+    {
+        let manager = globals.bind(qh, 1..=3, GlobalData)?;
+        Ok(Self { manager })
+    }
+
+    pub fn data_device_manager(&self) -> &WlDataDeviceManager {
+        &self.manager
+    }
+
+    /// creates a data source for copy paste
+    pub fn create_copy_paste_source<'s, D, I>(
+        &self,
+        qh: &QueueHandle<D>,
+        mime_types: I,
+    ) -> CopyPasteSource
+    where
+        D: Dispatch<WlDataSource, DataSourceData> + 'static,
+        I: IntoIterator<Item = &'s str>,
+    {
+        CopyPasteSource { inner: self.create_data_source(qh, mime_types, None) }
+    }
+
+    /// creates a data source for drag and drop
+    pub fn create_drag_and_drop_source<'s, D, I>(
+        &self,
+        qh: &QueueHandle<D>,
+        mime_types: I,
+        dnd_actions: DndAction,
+    ) -> DragSource
+    where
+        D: Dispatch<WlDataSource, DataSourceData> + 'static,
+        I: IntoIterator<Item = &'s str>,
+    {
+        DragSource { inner: self.create_data_source(qh, mime_types, Some(dnd_actions)) }
+    }
+
+    /// creates a data source
+    fn create_data_source<'s, D, I>(
+        &self,
+        qh: &QueueHandle<D>,
+        mime_types: I,
+        dnd_actions: Option<DndAction>,
+    ) -> WlDataSource
+    where
+        D: Dispatch<WlDataSource, DataSourceData> + 'static,
+        I: IntoIterator<Item = &'s str>,
+    {
+        let source = self.create_data_source_with_data(qh, Default::default());
+
+        for mime in mime_types {
+            source.offer(mime.to_string());
+        }
+        if let Some(dnd_actions) = dnd_actions {
+            source.set_actions(dnd_actions);
```

Comment:
Apparently `set_actions()` was added in version 3 of the protocol, while this code is binding version `1..=3`. So presumably this could result in a protocol error on a compositor implementing version 1 or 2.